jlm/mlir/backend/mlirgen.cpp

```diff
@@ -0,0 +1,269 @@
+ * See COPYING for terms of redistribution.
+ */
+
+#include "jlm/mlir/backend/mlirgen.hpp"
+
+#ifdef MLIR
+
+#include <jlm/rvsdg/bitstring/comparison.hpp>
+#include <jlm/rvsdg/bitstring/constant.hpp>
+#include <jlm/rvsdg/traverser.hpp>
+#include <jlm/rvsdg/node.hpp>
+
+//#include <jlm/llvm/ir/operators/GetElementPtr.hpp>
+//#include <jlm/llvm/ir/operators/load.hpp>
+//#include <jlm/llvm/ir/operators/operators.hpp>
+//#include <jlm/llvm/ir/operators/sext.hpp>
+//#include <jlm/llvm/ir/operators/store.hpp>
+//#include <jlm/rvsdg/bitstring/type.hpp>
+
+#include "llvm/Support/raw_os_ostream.h"
+#include "mlir/IR/Verifier.h"
+
+namespace jlm {
+namespace rvsdgmlir {
+
+void
+MLIRGen::print(mlir::rvsdg::OmegaNode &omega, const util::filepath &filePath)
+{
+  // Verify the module
+  if (failed(mlir::verify(omega))) {
+    omega.emitError("module verification error");
+    throw util::error("Verification of RVSDG-MLIR failed");
+  }
+  // Print the module
+  if (filePath == "") {
+    ::llvm::raw_os_ostream os(std::cout);
+    omega.print(os);
+  } else {
+    std::error_code ec;
+    ::llvm::raw_fd_ostream os(filePath.to_str(), ec);
+    omega.print(os);
+  }
+}
+
+mlir::rvsdg::OmegaNode
+MLIRGen::convertModule(const llvm::RvsdgModule &rvsdgModule)
+{
+  auto &graph = rvsdgModule.Rvsdg();
+
+  // TODO
+  // Add assert checking that the graph consists of a single omega node
+
+  return convertOmega(graph);
+}
+
+mlir::rvsdg::OmegaNode
+MLIRGen::convertOmega(const rvsdg::graph &graph)
+{
+  // Create the MLIR omega node
+  mlir::rvsdg::OmegaNode omega = builder->create<mlir::rvsdg::OmegaNode>(builder->getUnknownLoc());
+
+  // Create a block for the region as this is currently not done automatically
+  mlir::Region &region = omega.getRegion();
+  mlir::Block *omegaBlock = new mlir::Block;
+  region.push_back(omegaBlock);
+
+  // Convert the region of the omega
+  auto subregion = graph.root();
+  ::llvm::SmallVector<mlir::Value> *regionResults = convertSubregion(subregion, omegaBlock);
+
+  // Handle the result of the omega
+  auto omegaResult = builder->create<mlir::rvsdg::OmegaResult>(builder->getUnknownLoc(), *regionResults);
+  omegaBlock->push_back(omegaResult);
+
+  return omega;
+}
+
+::llvm::SmallVector<mlir::Value> *
+MLIRGen::convertSubregion(rvsdg::region *region, mlir::Block *block)
+{
+  // Handle arguments of the region
+  for (size_t i = 0; i < region->narguments(); ++i) {
+    auto type = convertType(&region->argument(i)->type());
+    block->addArgument(type, builder->getUnknownLoc());
+  }
+
+  // Create an MLIR node for each RVSDG node and store each pair in a
+  // hash map for easy lookup of corresponding MLIR nodes
+  std::unordered_map<rvsdg::node *, mlir::Value> nodes;
+  for (rvsdg::node *rvsdgNode: rvsdg::topdown_traverser(region)) {
+    nodes[rvsdgNode] = convertNode(rvsdgNode, block);
+  }
+
+  // Handle results of the region
+  ::llvm::SmallVector<mlir::Value> *results = new ::llvm::SmallVector<mlir::Value>();
```

Comment:
This will result in memory leaks, no? These allocated vectors are never deallocated again. Should this method not return: `::llvm::SmallVector<mlir::Value>`

Reply:
Not using pointer any more